### DISCUSSION TITLE
Downloading through proxy with authentication

### DISCUSSION THREAD

**Discussion by hgouveia** [hgouveia/node-downloader-helper#85] [General]
follow up discussion from issue #81

**Comment by hgouveia** [comment]
> Hello
> 
> I just did a test by removing all my code from my main.js of electron, and added only code from the clean electron project, and it still not working (I didn't change anything in package or npm install) but the new main.js is loading nothing but the the node-downloader and electron stuff.
> 
> for me, electron must load nothing if its not imported or required, npm install itself can change things in electron?
> 
> I am not using node-fetch, just https-proxy-agent and axios wich can touch the https module.
> 
> Thank you for helping

@khkassem it shouldn't load anything if not required, so the package.json shouldnt affect, but you could try to remove it just in case,  i also suggest to upgrade to a newer version of Electron into your project and see what happens

**Comment by khkassem** [comment]
Hello,

Finaly I figured it out, and you're not going to believe what it was, I have two problem, the first was the url with parameters which you solved last time. And the real proxy problem was that my https-proxy-agent package was not installed correctly, I have the package in package-lock.json but not in package.json - I dont know why ! so the run was not giving me any error when i am using https-proxy-agent but its not working correctly.

Thank you for your help

**Reply by hgouveia** [reply]
> Hello,
> 
> Finaly I figured it out, and you're not going to believe what it was, I have two problem, the first was the url with parameters which you solved last time. And the real proxy problem was that my https-proxy-agent package was not installed correctly, I have the package in package-lock.json but not in package.json - I dont know why ! so the run was not giving me any error when i am using https-proxy-agent but its not working correctly.
> 
> Thank you for your help

nice! i am glad to hear you were able to solved it!, dont hesitate to open new issue or discussion if you find any issue in the future

**Reply by Chaphasilor** [reply]
> my https-proxy-agent package was not installed correctly, I have the package in package-lock.json but not in package.json - I dont know why ! 

tip: always use `npm ci` if you want to re-create an issue, it installs packages from the lock file :)

**Comment by t800hk** [comment]
Hi @hgouveia,

I try to use sonarqube-scanner npm package via npx sitting behind a corporate proxy.
I can see that sonarqube-scanner@2.8.1 working just fine, it is able to download  https://binaries.sonarsource.com/Distribution/sonar-scanner-cli/sonar-scanner-cli-4.5.0.2216-linux.zip and execute sonarqube scan successfully.

However latest version sonarqube-scanner@2.8.2 uses node-downloader-helper fails to download: ERROR: impossible to download and extract binary: connect EHOSTUNREACH <some cloudfron.net ip>:443 .

According to https://github.com/bellingard/sonar-scanner-npm/commit/d9055b86ba260ca578758b051df41ba02e1006f1 they recently changed from download package to node-downloader-helper package. This causes the issue.

My question is: is there any way to force node-downloader-helper to use http_proxy / https_proxy environment variables? Or sonarqube-scanner team should enhance their code to support corporate proxy with node-downloader-helper package?

What do you think?

Thx,
Zoli

**Reply by hgouveia** [reply]
Hello @t800hk ,  I just checked "download" npm package, and it seems they used a package to handle http request called "got" , after checking "got" they also used similar approach as me where you need to assign a http proxy agent, so i dont really know how it worked before  but is something i will keep investigating later would be nice to suppot it out of the box, 

neverless you could try to use something like`http.globalAgent` or https  https://nodejs.org/api/http.html#httpglobalagent   to override any http/https request with proxy , or check this package https://www.npmjs.com/package/global-agent  , i dont have experience with that library but give it a try and let me know if worked for you

**Reply by t800hk** [reply]
Thx for the hints, I managed to implement a workaround (until the code is enhanced):

npm i global-agent
export GLOBAL_AGENT_HTTP_PROXY=http://mycorporateproxyhost:mycorporateproxyport
npx -n "-r global-agent/bootstrap" sonarqube-scanner ...

(btw: above npx command is actually executed from a jenkins pipeline, this way I only have to amend the build job)
